### PR TITLE
speed up epoch processing by caching shuffling results

### PR DESCRIPTION
**Pull request by tersec**

### CODE REVIEW

File: beacon_chain/spec/validator.nim [hunk 1]
```diff
@@ -75,21 +74,31 @@ func get_shuffled_seq*(seed: Eth2Digest,
...
   ## This function is factored to facilitate testing with
   ## https://github.com/ethereum/eth2.0-test-generators/tree/master/permutated_index
   ## test vectors, which the split of get_shuffling obfuscates.
+  let list_size = validators.len.uint64
 
   let
     active_validator_indices = get_active_validator_indices(validators, epoch)
     committees_per_epoch = get_epoch_committee_count(
       len(active_validator_indices)).int
+    # Both mapIt-type-conversions are an SSZ artifact. TODO remove.
```
Comment (arnetheduck) [inline beacon_chain/spec/validator.nim]: what if we passed in a var ShufflingCache and updated it here, on demand?

File: beacon_chain/spec/datatypes.nim [hunk 2]
```diff
@@ -417,6 +417,9 @@ type
     eth1_data_votes*: seq[Eth1DataVote]
     deposit_index*: uint64
 
+    # Not in spec. TODO: don't serialize or deserialize this.
+    shuffling_cache*: ShufflingCache
```
[the remark this answers is not in the document]
Comment (tersec) [reply beacon_chain/spec/datatypes.nim]: Probably, yes.

File: beacon_chain/spec/validator.nim (hunk 1 above)
Comment (tersec) [reply beacon_chain/spec/validator.nim]: Something along those lines was my initial design. It has advantages and disadvantages, including:

- It centralizes and encapsulates the cache abstraction more effectively than the current approach, which scatters it around the codebase a little, especially in genesis state creation and state transition functions.

- However, the current approach cleanly separates mutable from immutable parts of the codebase. I quite like this aspect and am inclined not to spread mutability too far, when cache access pattern is extremely predictable.

So, I can see arguments for both, but it's already getting basically 100% hit rates for me with a very simple algorithm that it's easier to reason about than a typical cache, so I'm not inclined to immediately switch this to a `var ShufflingCache` in the shuffling/get crosslinks area immediately.

Comment (arnetheduck) [reply beacon_chain/spec/validator.nim]: ok. I predict a refactoring, but that can be said of all the state transition code ;)

**Pull request merged**